File: grind75/sort_colors.go
package grind75

func sortColors(nums []int) {
	i, left, right := 0, 0, len(nums)-1
	for i <= right {
		if nums[i] == 0 {
			swap(nums, i, left)
			i++
			left++
		} else if nums[i] == 2 {
			swap(nums, i, right)
			right--
		} else {
			i++
		}
	}
}
func swap(nums []int, a, b int) {
	tmp := nums[a]
	nums[a] = nums[b]
	nums[b] = tmp
}

File: grind75/SortColors.java
package grind75;

public class SortColors {
    public void sortColors(int[] nums) {
        int i = 0, left = 0, right = nums.length-1;
        while (i <= right) { // only has 0, 1, 2 so enum element value and swap them
            if (nums[i] == 0) {
                swap(nums, i, left);
                left++;
                i++;
            } else if (nums[i] == 2) {
                swap(nums,i, right);
                right--;
            } else {
                i++;
            }
        }
    }    

    private void swap(int[] nums, int a, int b) {
        int tmp = nums[a];
        nums[a] = nums[b];
        nums[b] = tmp;
    }
}

